File: fsharp/two-bucket/TwoBucket.fs
module TwoBucket

type Bucket = One | Two

type BucketResult = {
    Moves: int 
    GoalBucket: Bucket 
    OtherBucket: int }

type State = int * int

type Move = 
    | FillOne
    | FillTwo
    | EmptyOne
    | EmptyTwo
    | PourRight
    | PourLeft

let allMoves = [FillOne; FillTwo; EmptyOne; EmptyTwo; PourRight; PourLeft]

let measure sizeOne sizeTwo (goal: int) (startBucket: Bucket) =
    let applyMove ((a, b): State) (move: Move)  = 
        match move with
        | FillOne   -> (sizeOne, b) 
        | FillTwo   -> (a, sizeTwo)
        | EmptyOne  -> (0, b)
        | EmptyTwo  -> (a, 0)
        | PourLeft  -> 
            let toPour = min (sizeOne - a) b 
            (a + toPour, b - toPour)
        | PourRight -> 
            let toPour = min (sizeTwo - b) a 
            (a - toPour, b + toPour)
    // tailored to be used with List.unfold
    let getNextStates ((states, explored): State list * State list) : (State list * (State list * State list)) option = 
        let nxtStates =
            states |> List.collect (fun st -> 
                allMoves 
                |> List.map (applyMove st)
                |> List.filter (fun ex -> not (List.contains ex explored)) 
            ) |> List.distinct
        if nxtStates = [] then 
            None 
        else
            Some (states, (nxtStates, explored |> List.append nxtStates))
    // put the initial state into getNextStates and unfold
    let allReachableStates : State list list =
        let start = if startBucket = Bucket.One then (sizeOne, 0) else (0, sizeTwo)
        let forbidden = if startBucket = Bucket.One then (0, sizeTwo) else (sizeOne, 0)
        List.unfold getNextStates ([start], [forbidden])
    // tailored to be used with List.choose to extract a BucketResult from a (idx, list of states) tuple
    let tryBucketResult (idx, ls) : BucketResult option = 
        if List.exists (fst >> (=) goal) ls  then 
            let (_, other) = List.find (fst >> (=) goal) ls 
            Some { Moves = idx + 1; GoalBucket = Bucket.One; OtherBucket = other }
        else if List.exists (snd >> (=) goal) ls then
            let (other, _) = List.find (snd >> (=) goal) ls 
            Some { Moves = idx + 1; GoalBucket = Bucket.Two; OtherBucket = other }
        else 
            None
    // use index of reachable state list to get number of necessary moves
    allReachableStates
    |> List.indexed 
    |> List.choose tryBucketResult
    |> List.head // the test cases here guarantee an existing solution, otherwise use tryHead and distinguish cases

File: csharp/two-bucket/TwoBucket.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Bucket { One, Two }

public class TwoBucketResult
{
    public TwoBucketResult(int moves, Bucket goalBucket, int otherBucket)
    {
        Moves = moves;
        GoalBucket = goalBucket;
        OtherBucket = otherBucket;
    }

    public int Moves { get; init; }
    public Bucket GoalBucket { get; init; }
    public int OtherBucket { get; init; }
}

public class TwoBucket
{
    private record struct State(int Fst, int Snd); // need value equality
    public TwoBucket(int bucketOne, int bucketTwo, Bucket startBucket)
    {
        sizeOne = bucketOne;
        sizeTwo = bucketTwo;
        if (startBucket == Bucket.One)
        {
            stateQueue.Enqueue((new State(sizeOne, 0), 1)); // last entry gives number of moves
            forbiddenStates.Add(new State(0, sizeTwo));
        }
        else
        {
            stateQueue.Enqueue((new State(0, sizeTwo), 1));
            forbiddenStates.Add(new State(sizeOne, 0));
        }
    }

    public TwoBucketResult Measure(int goal)
    {   // handle impossible cases
        if (goal > Math.Max(sizeOne, sizeTwo) || goal % GCD(sizeOne, sizeTwo) != 0) 
            throw new ArgumentException("No solution");
        return Search(goal);
    }

    private TwoBucketResult Search(int goal)
    {
        var (state, moves) = stateQueue.Dequeue();
        var (a, b) = state;
        if (a == goal || b == goal) 
            return new TwoBucketResult(moves, (a == goal) ? Bucket.One : Bucket.Two, (a == goal) ? b : a);
        var nextStates = GetNextStates(state).Except(forbiddenStates).ToList();
        foreach (var newState in nextStates) // update hash set and queue
            forbiddenStates.Add(newState);
        foreach (var newState in nextStates) 
            stateQueue.Enqueue((newState, moves + 1)); // one move more needed to get to this state
        return Search(goal); // recurse
    }
    private IEnumerable<State> GetNextStates(State state)
    {
        var (a, b) = state;
        var pourLeft = Math.Min(sizeOne - a, b);
        var pourRight = Math.Min(sizeTwo - b, a);
        return new List<State>
        {
            new State(sizeOne, b), 
            new State(a, sizeTwo),                             
            new State(0, b), 
            new State(a, 0),                                             
            new State(a + pourLeft, b - pourLeft), 
            new State(a - pourRight, b + pourRight) 
        }.Distinct();
    }

    private static int GCD(int a, int b) => b == 0 ? a : GCD(b, a % b);

    private int sizeOne;
    private int sizeTwo;
    private HashSet<State> forbiddenStates = new HashSet<State>();
    private Queue<(State, int)> stateQueue = new Queue<(State, int)>();
}
